### PR TITLE
chore(client-tunnel): hook up placeholder IPC server for Linux

### PR DESCRIPTION
**Pull request by ReactorScram**
Run the Linux Client or firezone-client-tunnel with `--act-as-tunnel` and it'll listen for incoming connections on a Unix Domain Socket.

### CODE REVIEW

File: rust/client-tunnel/src/lib.rs
Comment (thomaseizinger) [inline rust/client-tunnel/src/lib.rs:42]: Why not call this something like `--standalone`? In either case, it will open a tunnel so with my user-hat on, I don't really understand what "not acting as a tunnel" would do.

File: rust/client-tunnel/src/lib.rs
Comment (thomaseizinger) [inline rust/client-tunnel/src/lib.rs:91]: What you want is a codec: https://docs.rs/tokio-util/latest/tokio_util/codec/index.html

That turns an `AsyncRead` into a `Stream` and allows you to read entire messages using `.next()`. It already comes with the codec you are implementing here: https://docs.rs/tokio-util/latest/tokio_util/codec/length_delimited/struct.LengthDelimitedCodec.html

File: rust/client-tunnel/src/lib.rs
Comment (thomaseizinger) [inline rust/client-tunnel/src/lib.rs:99]: If you are using JSON, you could also use a newline codec: https://docs.rs/tokio-util/latest/tokio_util/codec/struct.LinesCodec.html

File: rust/client-tunnel/src/lib.rs
Comment (thomaseizinger) [reply rust/client-tunnel/src/lib.rs:42]: Alternatively, this could also be modeled using two sub-commands:

```
./client-tunnel standalone -u wss://api.firezone.dev ...
```
and
```
./client-tunnel daemon -u wss://api.firezone.dev ...
```

File: rust/client-tunnel/src/linux.rs
Comment (thomaseizinger) [inline rust/client-tunnel/src/linux.rs:191]: Nice, this is pretty cool! I had no idea unix sockets can do that.

File: rust/client-tunnel/src/linux.rs
Comment (thomaseizinger) [inline rust/client-tunnel/src/linux.rs:200]: Are we testing this in CI already? I am not aware of that. Why don't we implement straight away what we want?

File: rust/client-tunnel/src/linux.rs
Comment (thomaseizinger) [inline rust/client-tunnel/src/linux.rs]: Perhaps log the UID and GID of the process that connected.

**Review by thomaseizinger — COMMENTED** [review]
Great start!

I think we should use a proper codec for the UDS, otherwise this will get messy and we have to write a bunch of tests to make sure the connection works when we can instead just use off-the-shelf stuff.

Also, whilst it is easy, I think we should rename this to `headless-client`.

**Review by thomaseizinger — APPROVED** [review]
I'll be offline tomorrow, approving now to unblock but I'd like to see the codec stuff changed. Can also be done in a follow-up PR though.

File: rust/client-tunnel/src/lib.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/lib.rs:42]: Yeah if we can break CLI compatibility, subcommands might be the best. That way there is no default behavior that can become part of the API, by default it says "hey tell me a subcommand"

File: rust/client-tunnel/src/lib.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/lib.rs:91]: Yeah I remember that came up for `subzone` and I couldn't figure out how to use it. I'll spend an hour or so trying again

File: rust/client-tunnel/src/linux.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/linux.rs:191]: Yeah! Not on Windows though 😨 So unless I'm missing something, the security will be different on Windows. Maybe I can get away with "Write this cookie into a file that only a user with the right permissions could possibly write"

File: rust/client-tunnel/src/linux.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/linux.rs:200]: I can try it but adding users to groups might require a new login shell, so I'm not sure how the script will look. I just didn't feel like messing with it in this PR yet.

File: rust/client-tunnel/src/linux.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/linux.rs]: Good idea. I'll log PID too, then the GUI can log its own PID to match up. I think I can get PID from the UDS, doesn't hurt anything.

**Comment by ReactorScram** [comment]
If it's cool, I'm gonna open 3 follow-up ~~PRs~~ issues before I merge this one. "going wide" usually means I don't have to wait on CI as long, and this first piece can merge by itself without staying as WIP too long.

File: rust/client-tunnel/src/linux.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/linux.rs:200]: #4513 is an issue to add that test as soon as I can

File: rust/client-tunnel/src/lib.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/lib.rs:91]: #4514 is an issue for me to do that refactor as soon as I can

File: rust/client-tunnel/src/lib.rs
Comment (ReactorScram) [reply rust/client-tunnel/src/lib.rs:42]: Will complete for #4515

**Comment by ReactorScram** [comment]
Turning it off and back on since CI is acting strange.

**Pull request merged**
